### PR TITLE
training: record the free per-step loss, batch metric behind an optional plot_frequency

### LINKED ISSUE #671: LoRA loss monitoring can substantially increase training time vs mflux-cv
I switched an existing Krea 2 Raw LoRA training workflow from `mflux-cv` 0.18.30 to `mflux` 0.19.0 and noticed a large performance difference with otherwise equivalent settings.

With `mflux-cv`, loss was recorded for every training iteration without an obvious performance issue. The test run was estimated at roughly 80 hours.

With `mflux` 0.19.0 and `plot_frequency: 1`, the same run was estimated at roughly 232 hours. Based on a quick agent-assisted analysis of the two codebases, each loss update now appears to calculate an averaged loss across a validation batch rather than use the "free" values obtained after each iteration. That work is much more expensive than recording the current training-step loss, especially when repeated after every iteration as occurs when plot_frequency is 1.

I'm not sure whether this is:

- Intentional behavior, with `plot_frequency` now expected to be much higher (What is the user expected to set that to, reasonably? Why does it vary by model in the examples?)
- A useful `mflux-cv` behavior that was not carried into the community version (the mflux-cv approach was noisy and maybe not super representative of the results, but it was fast)
- A performance issue in the current averaged-loss implementation (seems unlikely but maybe)

Another side-effect of making the plot_frequency higher is that it takes longer before we get a loss.html file for display.

The averaged validation loss may be more representative, so I'm not necessarily suggesting reverting it. I just wanted to flag the unexpectedly large training-time impact and get some input into what the situation is WRT LoRA training loss calculation right now. Particularly since some official examples use `plot_frequency: 1` and that has a major training time impact.

### PR DESCRIPTION
**Pull request by fxd0h**
## What

Fixes #671, with the shape agreed there: the training loss the step already computed is now recorded every step for free, the batch metric became a second series behind `plot_frequency`, and `plot_frequency` is optional with a default of 20.

- The loop captured its loss and discarded it unrecorded, then paid up to 10 extra forwards per plot tick for a smoothed re-read of training samples. The free value now feeds the primary series, every step, and `loss.html` refreshes per step.
- The batch metric stays for whoever wants the smoother curve: same computation, its own series, rendered as a second line with a toggle, at `plot_frequency` ticks. The initial pre-training point stays too.
- `plot_frequency` was required; it now defaults to 20 (both ernie examples already shipped that). The z-image and flux2 examples shipped 1 and are now 20.
- Checkpoint statistics moved from a flat list to `{"step_loss": [...], "batch_loss": [...]}`. A legacy checkpoint loads its single series as the batch metric, which is what those entries were.

## Checklist (definition of done)

- [x] Tests added/updated run in CI by default: statistics save/load for both series, legacy checkpoint load through `from_spec`, and the `plot_frequency` default and explicit-value paths.
- [x] `ruff check` and `ruff format` are clean.
- [x] `CHANGELOG.md`: entry under `Unreleased` (follow-up commit referencing this PR number).
- [x] Docs updated where behavior changed: the three example configs.
- [ ] New model.
- [ ] New/changed CLI.

## Verification

Krea 2 Raw QLoRA on an M5 Max (q8 base, 512px, 10-image dataset, rank 16, 10 steps, same seed):

| | s/step |
|---|---|
| before, plotting effectively off | 5.41 |
| before, `plot_frequency: 1` | 17.47 |
| after, `plot_frequency` omitted (default 20) | 5.43 |

The per-step recording and per-step `loss.html` refresh cost nothing measurable. Inspected the run's artifacts: `loss.html` carries the 10-point step series plus the initial batch point, and the checkpoint's loss file has both keys.

Full suite with the CI selector: 1475 passed.

<details open><summary><h3>Greptile Summary</h3></summary>

This PR separates training statistics into per-step and periodic batch-loss series while making `plot_frequency` optional with a default of 20.
- Records the already-computed training loss after every finite monitored step.
- Preserves periodic batch-loss measurements and legacy checkpoint compatibility.
- Updates the HTML plot to render and toggle both series.
- Updates example configurations, documentation, and statistics tests.
</details>

<h3>Confidence Score: 5/5</h3>

The PR appears safe to merge, with no concrete changed-code failure identified.

The new metric lifecycle follows existing micro-batch step semantics, preserves legacy checkpoint data in the appropriate batch series, and handles empty or mixed series in the plotter.

<details open><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| src/mflux/models/common/training/trainer.py | Records each finite micro-batch loss and retains the frequency-gated batch metric without introducing an additional MLX synchronization. |
| src/mflux/models/common/training/statistics/statistics.py | Introduces separate persisted step and batch series while mapping legacy flat checkpoints to the batch series. |
| src/mflux/models/common/training/statistics/plotter.py | Supports plotting either or both loss series and adds a batch-series visibility toggle. |
| src/mflux/models/common/training/state/training_spec.py | Defaults omitted monitoring plot frequency to 20 while retaining positive-value validation. |
| tests/training/test_loss_series.py | Covers serialization, legacy loading, and default or explicit plot-frequency behavior. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A[Training micro-batch] --> B[Compute loss and gradients]
    B --> C{Loss finite?}
    C -- No --> D[Reset accumulation and skip]
    C -- Yes --> E[Record step-loss point]
    E --> F{plot_frequency reached?}
    F -- Yes --> G[Compute and record batch-loss point]
    F -- No --> H[Refresh loss plot]
    G --> H
    H --> I{Checkpoint due?}
    I -- Yes --> J[Save both loss series]
    I -- No --> A
    J --> A
```
</details>

<sub>Reviews (1): Last reviewed commit: ["training: record the free per-step loss,..."](https://github.com/mflux-community/mflux/commit/f2e78519be4ecb3ffd527e8a781f6e18c7cb702e) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=56852500)</sub>

## Summary by CodeRabbit

* **New Features**
  * Training charts now distinguish step loss from batch loss, with an optional Batch visibility toggle.
  * Monitoring plot frequency defaults to every 20 steps and can be customized.
  * Existing training checkpoints remain compatible with the updated loss format.

* **Bug Fixes**
  * Reduced the overhead of training-loss monitoring by limiting plot generation frequency.

### CODE REVIEW

**Review by ianscrivener — APPROVED** [review]
<sub>Approved in [CodeRabbit Change Stack](https://app.coderabbit.ai/change-stack/mflux-community/mflux/pull/672)</sub>
